File: p2bin.c
/* p2bin.c */
/*****************************************************************************/
/* SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only                     */
/*                                                                           */
/* AS-Portierung                                                             */
/*                                                                           */
/* Umwandlung von AS-Codefiles in Binaerfiles                                */
/*                                                                           */
/*****************************************************************************/

#include "stdinc.h"
#include <string.h>
#include <ctype.h>

#include "version.h"
#include "endian.h"
#include "bpemu.h"
#include "strutil.h"
#include "nls.h"
#include "nlmessages.h"
#include "p2bin.rsc"
#include "ioerrs.h"
#include "chunks.h"
#include "stringlists.h"
#include "cmdarg.h"
#include "toolutils.h"

#define BinSuffix ".bin"


typedef void (*ProcessProc)(
#ifdef __PROTOS__
const char *FileName, LongWord Offset
#endif
);


static CMDProcessed ParUnprocessed;

static FILE *TargFile;
static String SrcName, TargName;

static LongWord StartAdr, StopAdr, EntryAdr, RealFileLen;
static LongWord MaxGran, Dummy;
static Boolean StartAuto, StopAuto, AutoErase, EntryAdrPresent;

static Byte FillVal, ValidSegment;
static Boolean DoCheckSum;

static Byte SizeDiv;
static LongWord ANDMask, ANDEq;
static ShortInt StartHeader;

static ChunkList UsedList;


#ifdef DEBUG
#define ChkIO(s) ChkIO_L(s, __LINE__)

static void ChkIO_L(char *s, int line)
{
  if (errno != 0)
  {
    fprintf(stderr, "%s %d\n", s, line);
    exit(3);
  }
}
#endif

static void ParamError(Boolean InEnv, char *Arg)
{
  fprintf(stderr, "%s%s\n%s\n", getmessage((InEnv)?Num_ErrMsgInvEnvParam:Num_ErrMsgInvParam), Arg, getmessage(Num_ErrMsgProgTerm));
  exit(1);
}

#define BufferSize 4096
static Byte Buffer[BufferSize];

static void OpenTarget(void)
{
  LongWord Rest, Trans, AHeader;

  TargFile = fopen(TargName, OPENWRMODE);
  if (!TargFile)
    ChkIO(TargName);
  RealFileLen = ((StopAdr - StartAdr + 1) * MaxGran) / SizeDiv;

  AHeader = abs(StartHeader);
  if (StartHeader != 0)
  {
    memset(Buffer, 0, AHeader);
    if (fwrite(Buffer, 1, abs(StartHeader), TargFile) != AHeader)
      ChkIO(TargName);
  }

  memset(Buffer, FillVal, BufferSize);

  Rest = RealFileLen;
  while (Rest != 0)
  {
    Trans = min(Rest, BufferSize);
    if (fwrite(Buffer, 1, Trans, TargFile) != Trans)
      ChkIO(TargName);
    Rest -= Trans;
  }
}

static void CloseTarget(void)
{
  LongWord z, AHeader;

  AHeader = abs(StartHeader);

  /* write entry address to file? */

  if ((EntryAdrPresent) && (StartHeader != 0))
  {
    LongWord bpos;

    rewind(TargFile);
    bpos = ((StartHeader > 0) ? 0 : -1 - StartHeader) << 3;
    for (z = 0; z < AHeader; z++)
    {
      Buffer[z] = (EntryAdr >> bpos) & 0xff;
      bpos += (StartHeader > 0) ? 8 : -8;
    }
    if (fwrite(Buffer, 1, AHeader, TargFile) != AHeader)
      ChkIO(TargName);
  }

  if (EOF == fclose(TargFile))
    ChkIO(TargName);

  /* compute checksum over file? */

  if (DoCheckSum)
  {
    LongWord Sum, Size, Rest, Trans, Read;

    TargFile = fopen(TargName, OPENUPMODE);
    if (!TargFile)
      ChkIO(TargName);
    if (fseek(TargFile, AHeader, SEEK_SET) == -1)
      ChkIO(TargName);
    Size = Rest = FileSize(TargFile) - AHeader - 1;

    Sum = 0;
    while (Rest > 0)
    {
      Trans = min(Rest, BufferSize);
      Rest -= Trans;
      Read = fread(Buffer, 1, Trans, TargFile);
      if (Read != Trans)
        ChkIO(TargName);
      for (z = 0; z < Trans; Sum += Buffer[z++]);
    }
    errno = 0;
    printf("%s%08lX\n", getmessage(Num_InfoMessChecksum), LoDWord(Sum));
    Buffer[0] = 0x100 - (Sum & 0xff);

    /* Some systems require fflush() between read & write operations.  And
       some other systems again garble the file pointer upon an fflush(): */

    fflush(TargFile);
    if (fseek(TargFile, AHeader + Size, SEEK_SET) == -1)
      ChkIO(TargName);
    if (fwrite(Buffer, 1, 1, TargFile) != 1)
      ChkIO(TargName);
    fflush(TargFile);

    if (fclose(TargFile) == EOF)
      ChkIO(TargName);
  }

  if (Magic != 0)
    unlink(TargName);
}

static void ProcessFile(const char *FileName, LongWord Offset)
{
  FILE *SrcFile;
  Word TestID;
  Byte InpHeader, InpCPU, InpSegment;
  LongWord InpStart, SumLen;
  Word InpLen, TransLen, ResLen;
  Boolean doit;
  LongWord ErgStart, ErgStop;
  LongInt NextPos;
  Word ErgLen = 0;
  Byte Gran;

  SrcFile = fopen(FileName, OPENRDMODE);
  if (!SrcFile)
    ChkIO(FileName);

  if (!Read2(SrcFile, &TestID))
    ChkIO(FileName);
  if (TestID != FileID)
    FormatError(FileName, getmessage(Num_FormatInvHeaderMsg));

  errno = 0;
  if (!QuietMode)
    printf("%s==>>%s", FileName, TargName);
  ChkIO(OutName);

  SumLen = 0;

  do
  {
    ReadRecordHeader(&InpHeader, &InpCPU, &InpSegment, &Gran, FileName, SrcFile);

    if (InpHeader == FileHeaderStartAdr)
    {
      if (!Read4(SrcFile, &ErgStart))
        ChkIO(FileName);
      if (!EntryAdrPresent)
      {
        EntryAdr = ErgStart;
        EntryAdrPresent = True;
      }
    }

    else if (InpHeader == FileHeaderDataRec)
    {
      if (!Read4(SrcFile, &InpStart))
        ChkIO(FileName);
      if (!Read2(SrcFile, &InpLen))
        ChkIO(FileName);

      NextPos = ftell(SrcFile) + InpLen;
      if (NextPos >= FileSize(SrcFile) - 1)
        FormatError(FileName, getmessage(Num_FormatInvRecordLenMsg));

      doit = (FilterOK(InpHeader) && (InpSegment == ValidSegment));

      if (doit)
      {
        InpStart += Offset;
        ErgStart = max(StartAdr, InpStart);
        ErgStop = min(StopAdr, InpStart + (InpLen/Gran) - 1);
        doit = (ErgStop >= ErgStart);
        if (doit)
        {
          ErgLen = (ErgStop + 1 - ErgStart) * Gran;
          if (AddChunk(&UsedList, ErgStart, ErgStop - ErgStart + 1, True))
          {
            errno = 0;
            fprintf(stderr, " %s\n", getmessage(Num_ErrMsgOverlap));
            ChkIO(OutName);
          }
        }
      }

      if (doit)
      {
        /* an Anfang interessierender Daten */

        if (fseek(SrcFile, (ErgStart - InpStart) * Gran, SEEK_CUR) == -1)
          ChkIO(FileName);

        /* in Zieldatei an passende Stelle */

        if (fseek(TargFile, (((ErgStart - StartAdr) * Gran)/SizeDiv) + abs(StartHeader), SEEK_SET) == -1)
          ChkIO(TargName);

        /* umkopieren */

        while (ErgLen > 0)
        {
          TransLen = min(BufferSize, ErgLen);
          if (fread(Buffer, 1, TransLen, SrcFile) != TransLen)
            ChkIO(FileName);
          if (SizeDiv == 1) ResLen = TransLen;
          else
          {
            LongWord Addr;

            ResLen = 0;
            for (Addr = 0; Addr < (LongWord)TransLen; Addr++)
              if (((ErgStart * Gran + Addr) & ANDMask) == ANDEq)
                Buffer[ResLen++] = Buffer[Addr];
          }
          if (fwrite(Buffer, 1, ResLen, TargFile) != ResLen)
            ChkIO(TargName);
          ErgLen -= TransLen;
          ErgStart += TransLen;
          SumLen += ResLen;
        }
      }
      if (fseek(SrcFile, NextPos, SEEK_SET) == -1)
        ChkIO(FileName);
    }
    else
      SkipRecord(InpHeader, FileName, SrcFile);
  }
  while (InpHeader != 0);

  if (!QuietMode)
  {
    errno = 0; printf("  ("); ChkIO(OutName);
    errno = 0; printf(Integ32Format, SumLen); ChkIO(OutName);
    errno = 0; printf(" %s)\n", getmessage((SumLen == 1) ? Num_Byte : Num_Bytes)); ChkIO(OutName);
  }
  if (!SumLen)
  {
    errno = 0;
    fputs(getmessage(Num_WarnEmptyFile), stdout);
    ChkIO(OutName);
  }

  if (fclose(SrcFile) == EOF)
    ChkIO(FileName);
}

static ProcessProc CurrProcessor;
static LongWord CurrOffset;

static void Callback(char *Name)
{
  CurrProcessor(Name, CurrOffset);
}

static void ProcessGroup(char *GroupName_O, ProcessProc Processor)
{
  String Ext, GroupName;

  CurrProcessor = Processor;
  strmaxcpy(GroupName, GroupName_O, STRINGSIZE);
  strmaxcpy(Ext, GroupName, STRINGSIZE);
  if (!RemoveOffset(GroupName, &CurrOffset))
    ParamError(False, Ext);
  AddSuffix(GroupName, STRINGSIZE, getmessage(Num_Suffix));

  if (!DirScan(GroupName, Callback))
    fprintf(stderr, "%s%s%s\n", getmessage(Num_ErrMsgNullMaskA), GroupName, getmessage(Num_ErrMsgNullMaskB));
}

static void MeasureFile(const char *FileName, LongWord Offset)
{
  FILE *f;
  Byte Header, CPU, Gran, Segment;
  Word Length, TestID;
  LongWord Adr, EndAdr;
  LongInt NextPos;

  f = fopen(FileName, OPENRDMODE);
  if (!f)
    ChkIO(FileName);

  if (!Read2(f, &TestID))
    ChkIO(FileName);
  if (TestID != FileMagic)
    FormatError(FileName, getmessage(Num_FormatInvHeaderMsg));

  do
  {
    ReadRecordHeader(&Header, &CPU, &Segment, &Gran, FileName, f);

    if (Header == FileHeaderDataRec)
    {
      if (!Read4(f, &Adr))
        ChkIO(FileName);
      if (!Read2(f, &Length))
        ChkIO(FileName);
      NextPos = ftell(f) + Length;
      if (NextPos > FileSize(f))
        FormatError(FileName, getmessage(Num_FormatInvRecordLenMsg));

      if (FilterOK(Header) && (Segment == ValidSegment))
      {
        Adr += Offset;
        EndAdr = Adr + (Length/Gran)-1;
        if (Gran > MaxGran)
          MaxGran = Gran;
        if (StartAuto)
          if (StartAdr > Adr)
            StartAdr = Adr;
        if (StopAuto)
          if (EndAdr > StopAdr)
            StopAdr = EndAdr;
      }

      fseek(f, NextPos, SEEK_SET);
    }
    else
      SkipRecord(Header, FileName, f);
  }
  while (Header != 0);

  if (fclose(f) == EOF)
    ChkIO(FileName);
}

static CMDResult CMD_AdrRange(Boolean Negate, const char *Arg)
{
  if (Negate)
  {
    StartAdr = 0; StopAdr = 0x7fff;
    return CMDOK;
  }
  else
    return CMD_Range(&StartAdr, &StopAdr,
                     &StartAuto, &StopAuto, Arg);
}

static CMDResult CMD_ByteMode(Boolean Negate, const char *pArg)
{
#define ByteModeCnt 9
  static const char *ByteModeStrings[ByteModeCnt] =
  {
    "ALL", "EVEN", "ODD", "BYTE0", "BYTE1", "BYTE2", "BYTE3", "WORD0", "WORD1"
  };
  static Byte ByteModeDivs[ByteModeCnt] =
  {
    1, 2, 2, 4, 4, 4, 4, 2, 2
  };
  static Byte ByteModeMasks[ByteModeCnt] =
  {
    0, 1, 1, 3, 3, 3, 3, 2, 2
  };
  static Byte ByteModeEqs[ByteModeCnt] =
  {
    0, 0, 1, 0, 1, 2, 3, 0, 2
  };

  int z;
  UNUSED(Negate);

  if (*pArg == '\0')
  {
    SizeDiv = 1;
    ANDEq = 0;
    ANDMask = 0;
    return CMDOK;
  }
  else
  {
    String Arg;

    strmaxcpy(Arg, pArg, STRINGSIZE);
    NLS_UpString(Arg);
    ANDEq = 0xff;
    for (z = 0; z < ByteModeCnt; z++)
      if (strcmp(Arg, ByteModeStrings[z]) == 0)
      {
        SizeDiv = ByteModeDivs[z];
        ANDMask = ByteModeMasks[z];
        ANDEq   = ByteModeEqs[z];
      }
    if (ANDEq == 0xff)
      return CMDErr;
    else
      return CMDArg;
  }
}

static CMDResult CMD_StartHeader(Boolean Negate, const char *Arg)
{
  Boolean err;
  ShortInt Sgn;

  if (Negate)
  {
    StartHeader = 0;
    return CMDOK;
  }
  else
  {
    Sgn = 1;
    if (*Arg == '\0')
      return CMDErr;
    switch (as_toupper(*Arg))
    {
      case 'B':
        Sgn = -1;
        /* fall-through */
      case 'L':
        Arg++;
    }
    StartHeader = ConstLongInt(Arg, &err, 10);
    if ((!err) || (StartHeader > 4))
      return CMDErr;
    StartHeader *= Sgn;
    return CMDArg;
  }
}	

static CMDResult CMD_EntryAdr(Boolean Negate, const char *Arg)
{
  Boolean err;

  if (Negate)
  {
    EntryAdrPresent = False;
    return CMDOK;
  }
  else
  {
    EntryAdr = ConstLongInt(Arg, &err, 10);
    if (err)
      EntryAdrPresent = True;
    return (err) ? CMDArg : CMDErr;
  }
}

static CMDResult CMD_FillVal(Boolean Negate, const char *Arg)
{
  Boolean err;
  UNUSED(Negate);

  FillVal = ConstLongInt(Arg, &err, 10);
  return err ? CMDArg : CMDErr;
}

static CMDResult CMD_CheckSum(Boolean Negate, const char *Arg)
{
  UNUSED(Arg);

  DoCheckSum = !Negate;
  return CMDOK;
}

static CMDResult CMD_AutoErase(Boolean Negate, const char *Arg)
{
  UNUSED(Arg);

  AutoErase = !Negate;
  return CMDOK;
}

static CMDResult CMD_ForceSegment(Boolean Negate,  const char *Arg)
{
  int z = addrspace_lookup(Arg);

  if (z >= SegCount)
    return CMDErr;

  if (!Negate)
    ValidSegment = z;
  else if (ValidSegment == z)
    ValidSegment = SegCode;

  return CMDArg;
}

#define P2BINParamCnt (sizeof(P2BINParams) / sizeof(*P2BINParams))
static CMDRec P2BINParams[] =
{
  { "f"        , CMD_FilterList },
  { "r"        , CMD_AdrRange },
  { "s"        , CMD_CheckSum },
  { "m"        , CMD_ByteMode },
  { "l"        , CMD_FillVal },
  { "e"        , CMD_EntryAdr },
  { "S"        , CMD_StartHeader },
  { "k"        , CMD_AutoErase },
  { "q"        , CMD_QuietMode },
  { "QUIET"    , CMD_QuietMode },
  { "SEGMENT"  , CMD_ForceSegment },
};

int main(int argc, char **argv)	
{
  int z;
  char *ph1, *ph2;
  String Ver;

  nls_init();
  if (!NLS_Initialize(&argc, argv))
    exit(4);

  endian_init();
  strutil_init();
  bpemu_init();
  nlmessages_init("p2bin.msg", *argv, MsgId1, MsgId2);
  ioerrs_init(*argv);
  chunks_init();
  cmdarg_init(*argv);
  toolutils_init(*argv);

  InitChunk(&UsedList);

  if (argc <= 1)
  {
    errno = 0;
    printf("%s%s%s\n", getmessage(Num_InfoMessHead1), GetEXEName(argv[0]), getmessage(Num_InfoMessHead2));
    ChkIO(OutName);
    for (ph1 = getmessage(Num_InfoMessHelp), ph2 = strchr(ph1, '\n'); ph2; ph1 = ph2 + 1, ph2 = strchr(ph1, '\n'))
    {
      *ph2 = '\0';
      printf("%s\n", ph1);
      *ph2 = '\n';
    }
    exit(1);
  }

  StartAdr = 0;
  StopAdr = 0x7fff;
  StartAuto = True;
  StopAuto = True;
  FillVal = 0xff;
  DoCheckSum = False;
  SizeDiv = 1;
  ANDEq = 0;
  EntryAdr = -1;
  EntryAdrPresent = False;
  AutoErase = False;
  StartHeader = 0;
  ValidSegment = SegCode;
  ProcessCMD(argc, argv, P2BINParams, P2BINParamCnt, ParUnprocessed, "P2BINCMD", ParamError);

  if (!QuietMode)
  {
    as_snprintf(Ver, sizeof(Ver), "P2BIN/C V%s", Version);
    WrCopyRight(Ver);
  }

  if (ProcessedEmpty(ParUnprocessed))
  {
    errno = 0;
    fprintf(stderr, "%s\n", getmessage(Num_ErrMsgTargMissing));
    ChkIO(OutName);
    exit(1);
  }

  z = argc - 1;
  while ((z > 0) && (!ParUnprocessed[z]))
    z--;
  strmaxcpy(TargName, argv[z], STRINGSIZE);
  if (!RemoveOffset(TargName, &Dummy))
    ParamError(False, argv[z]);
  ParUnprocessed[z] = False;
  if (ProcessedEmpty(ParUnprocessed))
  {
    strmaxcpy(SrcName, argv[z], STRINGSIZE);
    DelSuffix(TargName);
  }
  AddSuffix(TargName, STRINGSIZE, BinSuffix);

  MaxGran = 1;
  if ((StartAuto) || (StopAuto))
  {
    if (StartAuto)
      StartAdr = 0xfffffffful;
    if (StopAuto)
      StopAdr = 0;
    if (ProcessedEmpty(ParUnprocessed))
      ProcessGroup(SrcName, MeasureFile);
    else
      for (z = 1; z < argc; z++)
        if (ParUnprocessed[z])
          ProcessGroup(argv[z], MeasureFile);
    if (StartAdr > StopAdr)
    {
      errno = 0;
      fprintf(stderr, "%s\n", getmessage(Num_ErrMsgAutoFailed));
      ChkIO(OutName);
      exit(1);
    }
    if (!QuietMode)
    {
      printf("%s: 0x%08lX-", getmessage(Num_InfoMessDeducedRange), LoDWord(StartAdr));
      printf("0x%08lX\n", LoDWord(StopAdr));
    }
  }

  OpenTarget();

  if (ProcessedEmpty(ParUnprocessed))
    ProcessGroup(SrcName, ProcessFile);
  else
    for (z = 1; z < argc; z++)
      if (ParUnprocessed[z])
        ProcessGroup(argv[z], ProcessFile);

  CloseTarget();

  if (AutoErase)
  {
    if (ProcessedEmpty(ParUnprocessed))
      ProcessGroup(SrcName, EraseFile);
    else
      for (z = 1; z < argc; z++)
        if (ParUnprocessed[z])
          ProcessGroup(argv[z], EraseFile);
  }

  return 0;
}

File: p2bin.pas
{$g-,i-,s+,r-,v-}
	PROGRAM ProgToBin;
        Uses Dos,
             StdHandl,Hex,NLS,Chunks,DecodeCM,
             AsmUtils;

CONST
   BinSuffix='.BIN';

TYPE
   ProcessProc=PROCEDURE(FileName:String; Offset:LongInt);

VAR
   ParProcessed:CMDProcessed;
   z,z2:Integer;

   TargFile:File;
   SrcName,TargName:String;

   StartAdr,StopAdr,RealFileLen:LongInt;
   MaxGran,Dummy:LongInt;
   StartAuto,StopAuto:Boolean;

   FillVal:Byte;
   DoCheckSum:Boolean;

   SizeDiv:Byte;
   ANDMask,ANDEq:LongInt;

   UsedList:ChunkList;

{$i IoErrors.rsc}
{$i Tools.rsc}
{$i P2bin.rsc}


        PROCEDURE ParamError(InEnv:Boolean; Arg:String);
        Far;
BEGIN
   IF InEnv THEN Write(ErrMsgInvEnvParam)
            ELSE Write(ErrMsgInvParam);
   WriteLn(Arg);
   WriteLn(ErrMsgProgTerm);
   Halt(1);
END;

        PROCEDURE OpenTarget;
CONST
   BufferSize=4096;
VAR
   Rest,Trans:LongInt;
   Buffer:ARRAY[0..BufferSize-1] OF Byte;
BEGIN
   Assign(TargFile,TargName); SetFileMode(2); Rewrite(TargFile,1);
   ChkIO(TargName); RealFileLen:=((StopAdr-StartAdr+1)*MaxGran) DIV SizeDiv;

   FillChar(Buffer,BufferSize,FillVal);

   Rest:=RealFileLen;
   WHILE Rest<>0 DO
    BEGIN
     Trans:=Min(Rest,BufferSize);
     BlockWrite(TargFile,Buffer,Trans); ChkIO(TargName);
     Dec(Rest,Trans);
    END;
END;

	PROCEDURE CloseTarget;
CONST
   BufferSize=4096;
VAR
   Sum,Rest,Trans,z:LongInt;
   Buffer:ARRAY[0..BufferSize-1] OF Byte;
BEGIN
   IF DoCheckSum THEN
    BEGIN
     Rest:=FileSize(TargFile);
     Seek(TargFile,0); ChkIO(TargName);
     Sum:=0;
     WHILE Rest<>0 DO
      BEGIN
       Trans:=Min(Rest,BufferSize);
       Dec(Rest,Trans);
       BlockRead(TargFile,Buffer,Trans); ChkIO(TargName);
       FOR z:=0 TO Pred(Trans) DO Inc(Sum,Buffer[z]);
       IF Rest=0 THEN Dec(Sum,Buffer[Pred(Trans)]);
      END;
     WriteLn(InfoMessChecksum,HexWord(Sum SHR 16),HexWord(Sum AND $ffff)); ChkIO(OutName);
     Seek(TargFile,FileSize(TargFile)-1); ChkIO(TargName);
     Buffer[0]:=$100-Lo(Sum);
     BlockWrite(TargFile,Buffer,1); ChkIO(TargName);
    END;

   Close(TargFile); ChkIO(TargName);
   IF Magic<>0 THEN Erase(TargFile);
END;

        PROCEDURE ProcessFile(FileName:String; Offset:LongInt);
        Far;
CONST
   BufferSize=4096;
VAR
   SrcFile:File;
   TestID:Word;
   InpHeader,InpSegment:Byte;
   InpStart,SumLen:LongInt;
   InpLen,TransLen,ResLen:Word;
   doit:Boolean;
   Buffer:ARRAY[0..BufferSize-1] OF Byte;
   ErgStart,ErgStop,NextPos:LongInt;
   ErgLen:Word;
   z:Integer;
   Gran:Byte;
BEGIN
   Assign(SrcFile,FileName); SetFileMode(0); Reset(SrcFile,1);
   ChkIO(FileName);

   BlockRead(SrcFile,TestID,2); ChkIO(FileName);
   IF TestID<>FileID THEN FormatError(FileName,FormatInvHeaderMsg);

   Write(FileName,'==>>',TargName); ChkIO(OutName);

   SumLen:=0;

   REPEAT
    ReadRecordHeader(InpHeader,InpSegment,Gran,FileName,SrcFile);
    IF InpHeader=FileHeaderStartAdr THEN
     BEGIN
      BlockRead(SrcFile,ErgStart,SizeOf(ErgStart)); ChkIO(FileName);
     END
    ELSE IF InpHeader<>FileHeaderEnd THEN
     BEGIN
      BlockRead(SrcFile,InpStart,4); ChkIO(FileName);
      BlockRead(SrcFile,InpLen,2); ChkIO(FileName);

      NextPos:=FilePos(SrcFile)+InpLen;
      IF NextPos>=FileSize(SrcFile)-1 THEN
       FormatError(FileName,FormatInvRecordLenMsg);

      doit:=FilterOK(InpHeader) AND (InpSegment=SegCode);

      IF doit THEN
       BEGIN
        Inc(InpStart,Offset);
	ErgStart:=Max(StartAdr,InpStart);
	ErgStop:=Min(StopAdr,InpStart+(InpLen DIV Gran)-1);
	doit:=IsUGreaterEq(ErgStop,ErgStart);
        IF doit THEN
	 BEGIN
	  ErgLen:=(ErgStop+1-ErgStart)*Gran;
          IF AddChunk(UsedList,ErgStart,ErgStop-ErgStart+1,True) THEN
           BEGIN
            WriteLn(' ',ErrMsgOverlap); ChkIO(OutName);
           END;
         END;
       END;

      IF doit THEN
       BEGIN
	{ an Anfang interessierender Daten }

	Seek(SrcFile,FilePos(SrcFile)+(ErgStart-InpStart)*Gran);

	{ in Zieldatei an passende Stelle }

	Seek(TargFile,(ErgStart-StartAdr)*Gran DIV SizeDiv);

	{ umkopieren }

	WHILE ErgLen>0 DO
	 BEGIN
	  TransLen:=Min(BufferSize,ErgLen);
          BlockRead(SrcFile,Buffer,TransLen); ChkIO(FileName);
	  IF SizeDiv=1 THEN ResLen:=TransLen
	  ELSE
	   BEGIN
	    ResLen:=0;
	    FOR z:=0 TO Pred(TransLen) DO
	    IF (ErgStart*Gran+z) AND ANDMask = ANDEq THEN
	     BEGIN
	      Buffer[ResLen]:=Buffer[z]; Inc(ResLen);
	     END;
	   END;
          BlockWrite(TargFile,Buffer,ResLen); ChkIO(TargName);
	  Dec(ErgLen,TransLen); Inc(ErgStart,TransLen); Inc(SumLen,ResLen);
	 END;
       END;
      Seek(SrcFile,NextPos); ChkIO(FileName);
     END;
   UNTIL InpHeader=0;

   WriteLn('  (',SumLen,' Byte)'); ChkIO(OutName);

   Close(SrcFile); ChkIO(FileName);
END;

        PROCEDURE MeasureFile(FileName:String; Offset:LongInt);
	Far;
VAR
   f:File;
   Header,Segment,Gran:Byte;
   Length,TestID:Word;
   Adr,EndAdr,HPos,NextPos:LongInt;
BEGIN
   Assign(f,FileName); SetFileMode(0); Reset(f,1);
   ChkIO(FileName);

   BlockRead(f,TestID,2); ChkIO(FileName);
   IF TestID<>FileMagic THEN FormatError(FileName,FormatInvHeaderMsg);

   REPEAT
    ReadRecordHeader(Header,Segment,Gran,FileName,f);

    IF Header=FileHeaderStartAdr THEN
     BEGIN
      Seek(f,FilePos(f)+SizeOf(LongInt)); ChkIO(FileName);
     END
    ELSE IF Header<>FileHeaderEnd THEN
     BEGIN
      BlockRead(f,Adr,4); ChkIO(FileName);
      BlockRead(f,Length,2); ChkIO(FileName);
      NextPos:=FilePos(f)+Length;
      IF NextPos>FileSize(f) THEN
       FormatError(FileName,FormatInvRecordLenMsg);

      IF FilterOK(Header) AND (Segment=SegCode) THEN
       BEGIN
        Inc(Adr,Offset);
        EndAdr:=Adr+(Length DIV Gran)-1;
        IF Gran>MaxGran THEN MaxGran:=Gran;
	IF StartAuto THEN
	 IF IsUGreater(StartAdr,Adr) THEN StartAdr:=Adr;
	IF StopAuto THEN
	 IF IsUGreater(EndAdr,StopAdr) THEN StopAdr:=EndAdr;
       END;

      Seek(f,NextPos);
     END;

   UNTIL Header=0;

   Close(f);
END;

        PROCEDURE ProcessGroup(GroupName:String; Processor:ProcessProc);
VAR
   s:SearchRec;
   Path,Name,Ext:String;
   Offset:LongInt;
BEGIN
   Ext:=GroupName;
   IF NOT RemoveOffset(GroupName,Offset) THEN ParamError(False,Ext);

   AddSuffix(GroupName,Suffix);

   FSplit(GroupName,Path,Name,Ext);

   FindFirst(GroupName,Archive,s);
   IF DosError<>0 THEN
    BEGIN
     WriteLn(ErrMsgNullMaskA,GroupName,ErrMsgNullMaskB);
     ChkIO(OutName);
    END
   ELSE
    WHILE DosError=0 DO
     BEGIN
      Processor(Path+s.Name,Offset);
      FindNext(s);
     END;
END;

	FUNCTION CMD_AdrRange(Negate:Boolean; Arg:String):CMDResult;
	Far;
VAR
   p,err:Integer;
BEGIN
   IF Negate THEN
    BEGIN
     StartAdr:=0; StopAdr:=$7fff;
     CMD_AdrRange:=CMDOK;
    END
   ELSE
    BEGIN
     CMD_AdrRange:=CMDErr;

     p:=Pos('-',Arg); IF p=0 THEN Exit;

     IF Copy(Arg,1,p-1)='$' THEN
      BEGIN
       StartAuto:=True; err:=0;
      END
     ELSE
      BEGIN
       StartAuto:=False;
       ULongVal(Copy(Arg,1,p-1),StartAdr,err);
      END;
     IF err<>0 THEN Exit;

     IF Copy(Arg,p+1,Length(Arg)-p)='$' THEN
      BEGIN
       StopAuto:=True; err:=0;
      END
     ELSE
      BEGIN
       StopAuto:=False;
       ULongVal(Copy(Arg,p+1,Length(Arg)-p),StopAdr,err);
      END;
     IF err<>0 THEN Exit;

     IF (NOT StartAuto) AND (NOT StopAuto) AND IsUGreater(StartAdr,StopAdr)
     THEN Exit;

     CMD_AdrRange:=CMDArg;
    END;
END;

	FUNCTION CMD_ByteMode(Negate:Boolean; Arg:String):CMDResult;
	Far;
CONST
   ByteModeCnt=9;
   ByteModeStrings:ARRAY[1..ByteModeCnt] OF String=('ALL','EVEN','ODD','BYTE0','BYTE1','BYTE2','BYTE3','WORD0','WORD1');
   ByteModeDivs   :ARRAY[1..ByteModeCnt] OF Byte  =(1,2,2,4,4,4,4,2,2);
   ByteModeMasks  :ARRAY[1..ByteModeCnt] OF Byte  =(0,1,1,3,3,3,3,2,2);
   ByteModeEqs    :ARRAY[1..ByteModeCnt] OF Byte  =(0,0,1,0,1,2,3,0,2);
VAR
   z:Integer;
BEGIN
   IF Arg='' THEN
    BEGIN
     SizeDiv:=1; ANDEq:=0; ANDMask:=0;
     CMD_ByteMode:=CMDOK;
    END
   ELSE
    BEGIN
     FOR z:=1 TO Length(Arg) DO Arg[z]:=UpCase(Arg[z]);
     ANDEq:=$ff;
     FOR z:=1 TO ByteModeCnt DO
     IF Arg=ByteModeStrings[z] THEN
      BEGIN
       SizeDiv:=ByteModeDivs[z];
       ANDMask:=ByteModeMasks[z];
       ANDEq  :=ByteModeEqs[z];
      END;
     IF AndEq=$ff THEN CMD_ByteMode:=CMDErr ELSE CMD_ByteMode:=CMDArg;
    END;
END;

	FUNCTION CMD_FillVal(Negate:Boolean; Arg:String):CMDResult;
	Far;
VAR
   err:Integer;
BEGIN
   Val(Arg,FillVal,err);
   IF err<>0 THEN CMD_FillVal:=CMDErr ELSE CMD_FillVal:=CMDArg;
END;

	FUNCTION CMD_CheckSum(Negate:Boolean; Arg:String):CMDResult;
	Far;
BEGIN
   DoCheckSum:=NOT Negate;
   CMD_CheckSum:=CMDOK;
END;

CONST
   P2BINParamCnt=5;
   P2BINParams:ARRAY[1..P2BINParamCnt] OF CMDRec=
	       ((Ident:'f'; Callback:CMD_FilterList),
		(Ident:'r'; Callback:CMD_AdrRange),
		(Ident:'s'; Callback:CMD_CheckSum),
		(Ident:'m'; Callback:CMD_ByteMode),
		(Ident:'l'; Callback:CMD_FillVal));

BEGIN
   NLS_Initialize; WrCopyRight('P2BIN/2 V1.41r6','P2BIN V1.41r6');

   InitChunk(UsedList);

   IF ParamCount=0 THEN
    BEGIN
     WriteLn(InfoMessHead1,GetEXEName,InfoMessHead2); ChkIO(OutName);
     FOR z:=1 TO InfoMessHelpCnt DO
      BEGIN
       WriteLn(InfoMessHelp[z]); ChkIO(OutName);
      END;
     Halt(1);
    END;

   StartAdr:=0; StopAdr:=$7fff; StartAuto:=False; StopAuto:=False;
   FillVal:=$ff; DoCheckSum:=False; SizeDiv:=1; AndEq:=0;
   ProcessCMD(@P2BINParams,P2BINParamCnt,ParProcessed,'P2BINCMD',ParamError);

   IF ParProcessed=[] THEN
    BEGIN
     WriteLn(ErrMsgTargMissing);
     ChkIO(OutName);
     Halt(1);
    END;

   z:=ParamCount;
   WHILE (z>0) AND (NOT (z IN ParProcessed)) DO Dec(z);
   TargName:=ParamStr(z);
   IF NOT RemoveOffset(TargName,Dummy) THEN ParamError(False,ParamStr(z));
   ParProcessed:=ParProcessed-[z];
   IF ParProcessed=[] THEN
    BEGIN
     SrcName:=ParamStr(z); DelSuffix(TargName);
    END;
   AddSuffix(TargName,BinSuffix);

   MaxGran:=1;
   IF StartAuto OR StopAuto THEN
    BEGIN
     IF StartAuto THEN StartAdr:=-1;
     IF StopAuto THEN StopAdr:=0;
     IF ParProcessed=[] THEN ProcessGroup(SrcName,MeasureFile)
     ELSE FOR z:=1 TO ParamCount DO
      IF z IN ParProcessed THEN ProcessGroup(ParamStr(z),MeasureFile);
     IF IsUGreater(StartAdr,StopAdr) THEN
      BEGIN
       WriteLn(ErrMsgAutoFailed); ChkIO(OutName); Halt(1);
      END;
    END;

   OpenTarget;

   IF ParProcessed=[] THEN ProcessGroup(SrcName,ProcessFile)
   ELSE FOR z:=1 TO ParamCount DO
    IF z IN ParProcessed THEN ProcessGroup(ParamStr(z),ProcessFile);


   CloseTarget;
END.
